File: src/main/java/com/uiptv/service/CacheServiceImpl.java
package com.uiptv.service;

import com.uiptv.api.LoggerCallback;
import com.uiptv.db.ChannelDb;
import com.uiptv.model.Account;
import com.uiptv.service.cache.AccountCacheReloaderFactory;
import com.uiptv.util.FetchAPI;

import java.io.IOException;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import static com.uiptv.model.Account.AccountAction.itv;

public class CacheServiceImpl implements CacheService {
    private final AccountCacheReloaderFactory reloaderFactory = new AccountCacheReloaderFactory();

    private static Map<String, String> getCategoryParams(Account.AccountAction accountAction) {
        final Map<String, String> params = new HashMap<>();
        params.put("JsHttpRequest", new Date().getTime() + "-xml");
        params.put("type", accountAction.name());
        params.put("action", accountAction == itv ? "get_genres" : "get_categories");
        return params;
    }

    @Override
    public void reloadCache(Account account, LoggerCallback logger) throws IOException {
        reloaderFactory.get(account.getType()).reloadCache(account, logger);
    }

    @Override
    public boolean verifyMacAddress(Account account, String macAddress) {
        if (account == null) {
            return false;
        }

        String originalMac = account.getMacAddress();
        try {
            account.setMacAddress(macAddress);
            HandshakeService.getInstance().connect(account);

            if (account.isNotConnected()) {
                return false;
            }

            String jsonCategories = FetchAPI.fetch(getCategoryParams(account.getAction()), account);
            return !CategoryService.getInstance().parseCategories(jsonCategories, false).isEmpty();
        } catch (Exception e) {
            return false;
        } finally {
            account.setMacAddress(originalMac);
        }
    }

    @Override
    public int getChannelCountForAccount(String accountId) {
        return ChannelDb.get().getChannelCountForAccount(accountId);
    }
}

File: server/src/main/kotlin/com/uiptv/service/CacheServiceImpl.kt
package com.uiptv.service

import com.uiptv.api.LoggerCallback
import com.uiptv.db.ChannelDb
import com.uiptv.model.Account
import com.uiptv.service.cache.AccountCacheReloaderFactory
import com.uiptv.util.AccountType
import java.util.Date
import java.util.HashMap

class CacheServiceImpl(
    private val handshakeServiceProvider: () -> HandshakeService,
    private val categoryServiceProvider: () -> CategoryService,
    private val configurationServiceProvider: () -> ConfigurationService,
    private val channelServiceProvider: () -> ChannelService,
    private val fetchProvider: (Map<String, String>, Account) -> String
) : CacheService {
    private val reloaderFactory by lazy(LazyThreadSafetyMode.NONE) {
        AccountCacheReloaderFactory(
            categoryServiceProvider = categoryServiceProvider,
            configurationServiceProvider = configurationServiceProvider,
            handshakeServiceProvider = handshakeServiceProvider,
            channelServiceProvider = channelServiceProvider,
            fetchProvider = fetchProvider
        )
    }

    override fun reloadCache(account: Account, logger: LoggerCallback) {
        reloaderFactory.get(account.type).reloadCache(account, logger)
    }

    override fun verifyMacAddress(account: Account?, macAddress: String?): Boolean {
        if (account == null || macAddress.isNullOrBlank() || account.type != AccountType.STALKER_PORTAL) {
            return false
        }
        val originalMac = account.macAddress
        return try {
            account.macAddress = macAddress
            handshakeServiceProvider.invoke().connect(account)
            if (account.isNotConnected()) {
                false
            } else {
                val jsonCategories = fetchProvider.invoke(getCategoryParams(account.action), account)
                categoryServiceProvider.invoke().parseCategories(jsonCategories, false).isNotEmpty()
            }
        } catch (_: Exception) {
            false
        } finally {
            account.macAddress = originalMac
        }
    }

    override fun getChannelCountForAccount(accountId: String): Int = ChannelDb.get().getChannelCountForAccount(accountId)

    private fun getCategoryParams(accountAction: Account.AccountAction): Map<String, String> {
        val params = HashMap<String, String>()
        params["JsHttpRequest"] = Date().time.toString() + "-xml"
        params["type"] = accountAction.name
        params["action"] = if (accountAction == Account.AccountAction.itv) "get_genres" else "get_categories"
        return params
    }

    companion object {
        @JvmField
        val INSTANCE: CacheService = run {
            val cacheRef = arrayOfNulls<CacheService>(1)
            val channelRef = arrayOfNulls<ChannelService>(1)
            cacheRef[0] = CacheServiceImpl(
                handshakeServiceProvider = { HandshakeService.INSTANCE },
                categoryServiceProvider = { CategoryService.INSTANCE },
                configurationServiceProvider = { ConfigurationService },
                channelServiceProvider = { channelRef[0]!! },
                fetchProvider = com.uiptv.util.FetchAPI::fetch
            )
            channelRef[0] = ChannelService(
                cacheServiceProvider = { cacheRef[0]!! },
                contentFilterService = ContentFilterService,
                logoResolverService = LogoResolverService,
                configurationService = ConfigurationService,
                handshakeService = HandshakeService.INSTANCE
            )
            cacheRef[0]!!
        }
    }
}
